### PR TITLE
mzcompose: rewrite command-line driver

### PR DESCRIPTION
**Pull request by benesch**
Rewrite the mzcompose CLI driver with saner argument handling. This
fixes a number of long-standing bugs where arguments in the wrong order
were not properly forwarded to Docker Compose. In particular, passing
`--help` is now guaranteed to get you the *right* help information
(i.e., mzcompose's, if it applies). The old code would sometimes fail to
give you the mzcompose-specific help.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

I recommend viewing this PR mostly as new code. It's probably easier just to check out the PR, driving mzcompose around a bit, and making sure it behaves as you expect, rather than reading the implementation in too much detail.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Review by guswynn — COMMENTED** [review]
I do love argparse, but it really requires a lot of work lol

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:420]: the argparse api is weird, do you not set it on the `Namespace` here instead?

Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:420]: also, instead of this, it may be better to use `nargs=argparse.REMAINDER`: https://stackoverflow.com/questions/22850332/getting-the-remaining-arguments-in-argparse/46250042

It doesn't look like any command distinguishes between `unknown_args` and `unknown_subargs`, so it might be cleaner, however, its been a while since I've used that, so ymmv

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:243]: i'd make `list` a shorthand for this, `add_parser` takes `aliases=List[str]`

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:237]: does this mean --help wont load the composition anymore? thats so nice

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:288]: I think here we need to also add help for the workflow itself. It's out of scope for the pr, but the `mzworkflows.py` should be able to add help (like explaining how `TD_TEST`/positional version of `TD_TEST` works, and we should print that here, maybe after the `docker-compose` help. I like the `replace` below, but a log line referencing that "this subcommand will use docker-compose to run {workflow_name} workflow" or whatever would be really nice! otherwise, for new people it can be a bit inscrutable what is actually happening (even below we `announce` we are delegating)

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:60]: if they pass these, do we want a runtime exception?

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:317]: random q, do we subprocess as root or is the docker daemon handling this stuff?

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:344]: if you `add_help = True` you could add this help during the `add_parser` call

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:354]: list-workflows?

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [inline misc/python/materialize/cli/mzcompose.py:384]: these `DockerCompose` commands could add help in their `add_parser` call saying they just delegate

**Review by guswynn — CHANGES_REQUESTED** [review]

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:60]: Yes! That happens on L99 (and matches what currently happens).

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:49]: Ack, yeah, I'd rather not have two ways to do things, but I'd endorse swapping out the `--build-mode` flag for `--dev` and `--release` everywhere!

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:237]: Yes exactly!

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:243]: That's a good idea. If you don't mind I'll address that in a follow-up PR though since I want to avoid making too many behavior changes in this mega PR.

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:288]: Yeah, that would that'd be awesome! `RunCommand` further overrides this help btw to note that it can run a workflow in addition to a service.

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:317]: Yeah it's the Docker daemon. The client runs as you, but connects to the Docker daemon which does arbitrary stuff (like making volumes) as root.

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py]: Oops, thanks!

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py]: Heh, I went down that road. `argparse.REMAINDER` was deprecated in Python 3.9 because it has unfixable bugs: http://bugs.python.org/issue17050. It also doesn't distinguish between unknown args before the subcommand and unknown args after the subcommand, which is extremely important for properly forwarding args to `docker-compose`.

Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py]: I added that as a comment though! 

> the argparse api is weird, do you not set it on the `Namespace` here instead?

You _do_, it's just that `args` becomes the namespace. I refactored to make it clearer. (Sorry, I copied this out of the Python standard library default impl of `parse_known_args`, which had that confusing naming bit.)

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:344]: So `add_help = True` adds an argparse managed help message, but we want to roll our own `--help` flag here (via `DockerComposeCommand`) that forwards the help through to Docker Compose.

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:384]: 👍🏽 I'll punt this one to a follow up too!

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:49]: don't need to do that in pr, but a todo would be nice!

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:60]: ah I missed that!

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:243]: I like to add TODOs so I am less likely to forget!

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:344]: Oh I see, you are also getting the forwarded docker-compose help

File: misc/python/materialize/cli/mzcompose.py
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:384]: sounds good!

**Review by guswynn — APPROVED** [review]
personally I would add TODOs to track all the little later things, but your call, other than that, this is MUCH better

**Comment by benesch** [comment]
Ok perfect! Thank you very much for the review! I’ll file an issue tracking
all the follow ups before I hit merge.

File: misc/python/materialize/cli/mzcompose.py
Comment (benesch) [reply misc/python/materialize/cli/mzcompose.py:454]: Actually knocked this one out now.

**Comment by benesch** [comment]
I pushed a few more changes to smooth out error handling. Filed #9796 for followups. Merging on green! 🤞🏽 Thanks again for the review.

**Pull request merged**
